src/ClassicUO.IO/PinnedBuffer.cs

```diff
@@ -0,0 +1,46 @@
+﻿// SPDX-License-Identifier: BSD-2-Clause
+// author: Max Kellermann <max.kellermann@gmail.com>
+
+using System;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace ClassicUO.IO;
+
+/// <summary>
+///     A buffer with a fixed address, i.e. the allocation is pinned.
+/// </summary>
+public unsafe class PinnedBuffer : IDisposable
```

Comment:
I am thinking this should be a struct.

Reply:
Then it cannot be IDisposable, can it?

Reply:
Ref structs still cannot yet have interfaces, but they are disposable with duct typing. However, regular structs can be disposable.

Reply:
I barely know C# - but from reading the docs (https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/struct) I don't see why PinnedBuffer should be a struct.
(I also cannot see any text there that allows structs to have a base class.)

Reply:
Okay, "a structure type can implement interfaces", but: "You can't declare a finalizer within a structure type" - this part is extremely important because a `struct PinnedBuffer` without a finalizer would leak memory.

Reply:
"There exist also boxing and unboxing conversions between a structure type and any interface that it implements"
This seems to imply that casting `struct PinnedBuffer` to `IDisposable` would allocate a new object, copying the existing `PinnedBuffer`. That doesn't sound like a good idea at all. It sounds like it opens the door for obscure memory leaks. I don't think anybody should ever make a struct implement IDisposable!

Reply:
Correct, whatever is holding it, would have to treat pinned buffer as an unmanaged resource and do something like:

```cs
buffer.Dispose();
buffer = default;
```

Using a struct will allow the JIT/PGO to inline and optimize the code better, which is very useful for reading. We want to effectively turn the data reading into a long procedural code chain at runtime. Since the PinnedBuffer will exist on loaders that will never be deallocated anyways (singletons), this should be fine.

Reply:
> "There exist also boxing and unboxing conversions between a structure type and any interface that it implements" This seems to imply that casting `struct PinnedBuffer` to `IDisposable` would allocate a new object, copying the existing `PinnedBuffer`. That doesn't sound like a good idea at all. It sounds like it opens the door for obscure memory leaks. I don't think anybody should ever make a struct implement IDisposable!

That is correct, there is boxing until the JIT/PGO optimizes away the boxing which would definitely happen in our use-case. Furthermore, you don't have to use the IDisposable interface at all. It can be done via duct typing for our use case which avoids boxing.

Reply:
What code shall be inlined and optimized? `PinnedBuffer` has no code; only very trivial getters which can be inlined, no matter if it's class or struct.
`PinnedBuffer` is, in its current form, a base class for others such as `UOFile`, and `UOFile` really "owns" the memory allocation. Again, I know little of C#, but I don't think it really works the way you want it to. I still don't understand why it should be a struct, and I doubt it's even possible.

Reply:
If you are planning on PinnedBuffer being a base class, then you are correct. I thought you wanted to simply use it as a field to store the pointer + length. In most cases, accessing a field on a class that is an object, requires getting and dereferencing a pointer under the hood. If you were simply using PinnedBuffer as a field on a class, then making it struct means the fields of the struct can be optimized and sometimes inlined in the class directly so there is no extra indirection/dereference.

Reply:
.NET Core is nothing like .NET Framework, and it doesn't follow the same principles anymore. Lots of the performance gain is through nuanced and obscure understanding of the JIT/PGO :(

Reply:
> Furthermore, you don't have to use the IDisposable interface at all. It can be done via duct typing for our use case which avoids boxing.

But having `IDisposable` is the whole point of `PinnedBuffer` - it manages ownership of the memory (or rather: the concrete subclass does it; the subclass allocates the memory and frees it in the `Dispose(bool)` override). If it weren't for that, we wouldn't need this class at all - we could simply use `ReadOnlySpan<byte>` instead.
I don't get your point.

Reply:
You can't use `ReadOnlySpan<byte>` as a field on a class because it is a ref struct. Otherwise I would have suggested we use ReadOnlySpan. You can use `Memory<byte>` and get back a `ref T` (managed version of a pointer). It's simply more work to get back the raw pointer compared to building your own equiv of Memory<T> that uses/exposes the pointer directly.

It is normal to have structs that are IDisposable, and to have ref structs that duct type with a Dispose() method. If you don't want to do that, that is fine. Just understand that accessing the fields of PinnedBuffer will sometimes be slower.

Reply:
> If you are planning on PinnedBuffer being a base class

This is not a plan - it is what this PR already is. It's done.
And it isn't even a new thing - it's just a new name for the  `class DataReader` that exists for years.

The whole discussion about whether to make it a struct or class is orthogonal to this PR, because this PR doesn't introduce something new here.

Reply:
I think what you are looking for is a lighter/more flexible version of the [MemoryManager<T>](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.memorymanager-1?view=net-8.0). Then yes, we should use a class.

Reply:
> > If you are planning on PinnedBuffer being a base class
> 
> This is not a plan - it is what this PR already is. It's done. And it isn't even a new thing - it's just a new name for the `class DataReader` that exists for years.
> 
> The whole discussion about whether to make it a struct or class is orthogonal to this PR, because this PR doesn't introduce something new here.

yeah sorry, I missed this part:

`public class UOFile : PinnedBuffer`

I understand what you are getting at now.

Reply:
> You can't use `ReadOnlySpan<byte>` as a field on a class because it is a ref struct

Oh, indeed! .NET is truly a horrible mess.

> Just understand that accessing the fields of PinnedBuffer will sometimes be slower.

I do understand that very well, but that is not something I consider worth optimizing. We're talking about chasing 1 additional pointer, once, for parsing a file that weighs megabytes. The rest of the function is a million or a billion times slower than that one pointer dereference.

Of course, pointers dereferencing should be kept out of hot inner loops. I did a lot to reduce exactly that, because I understand that's where small optimizations really count.